### ISSUE TITLE
whatshap input pipeline for multi-alleleic testing

### ISSUE THREAD

**Issue by ekg** [vgteam/vg#565]
@shilpagarg has extend the WhatsHap algorithm to handle more than two candidate alleles.

We want to see how this works for haplotyping in assembly graphs.

The goal is to take an assembly graph made from illumina reads and then use pacbio reads to establish long-range phasing by considering bubbles in the assembly graph that are traversed by pacbio reads.

We would like to simplify things by considering only paths (alleles) through the bubbles that are supported by the pacbio reads. We want to consider only some top N candidate alleles. It looks like we can get what we need out of `vg genotype`, provided we have removed non-match edits from the pacbio reads using a slightly modified version of `vg scrub`. We then take the Locus records emitted by `vg genotype` and pass them into `vg locify` along with the scrubbed alignments so as to make the correct output for the graph-enabled version of WhatsHap. We want to consider only the bubbles that have more than 1 candidate allele and we would like to only keep the N best alleles, where N is a small number like 4 or 5.

Here are some blocking issues.

1) The genotyper can only be run effectively on small regions of the graph. We would like to be able to run it over each bubble in the base graph in parallel, as in this application we are only considering bubbles in the assembly graph that we have used as the reference.

2) `vg scrub` doesn't provide precisely the right behavior. We would like to remove any non-match edits from a set of alignments and replace them with match edits that are appropriate. In a sense we would like to use the assembly graph as a filter on the reads so that we can then consider only the phasing between successive bubbles, not the new edits or errors in the pacbio or illumina reads that we align to the the vg graph generated from the assembly graph.

3) We would like to be able to easily remove the lowest-supported alleles from a stream of Locus records. This could be implemented in a new command line utility (with filtering functions operating on Locus recards) or maybe directly in `vg locify`.

**Comment by ekg** [comment]
After discussing this with @edawson and @shilpagarg we came to the conclusion that the changes to scrub are complicated and not really required. It will be easiest to do things in vg locify following this pattern.

We make the assembly graph from the illumina data, then align the original illumina reads back to it. We use these alignments and the graph in `vg genotype` to generate loci calls. We then align pacbio reads to the graph and finally use these pacbio reads, the graph, and the loci as input to `vg locify`. We then just need to modify `vg locify` to take only the top N alleles at each locus, sorted by support in the reads.

**Comment by shilpagarg** [comment]
Just to make clear, top N alleles at each locus, sorted by support in the
pacbio reads. The reason to do vg genotype based on illumina alignments
only is because vg genotype is not yet designed to handle PacBio reads.
The output of vg locify is the pruned and sorted LOCUS and GAM records to
these top-N paths, where n is user-defined.

**Comment by edawson** [comment]
> we came to the conclusion that the changes to scrub are complicated and not really required

Specifically, there is ambiguity about what to replace non-match alignments with when scrubbing. Do we transform them to 'reference' matches, or the nearest allele by edit distance, or do we simply create a new alignment that has a noncontinuous path containing only match edits? It's a bit moot for the moment since an adjustment to the workflow solves the current issue, but I think it is an interesting question.

**Comment by faithokamoto** [comment]
Looks like the pipelines here use workflows that we've since deprecated and removed (vg scrub, vg locify). If we want to work on this or a related problem, a new Issue with updated vg should be opened.
